File: bin/record_artifact_hashes.py
from __future__ import annotations

import argparse
import csv
import hashlib
import json
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path


@dataclass(frozen=True)
class FileHash:
    path: str
    size_bytes: int
    mtime_utc: str
    sha256: str


def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(1024 * 1024), b""):
            h.update(chunk)
    return h.hexdigest()


def iso_utc_from_timestamp(ts: float) -> str:
    return datetime.fromtimestamp(ts, tz=timezone.utc).isoformat()


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Record local hashes for manifest-tracked artifacts under data/artifacts."
    )
    parser.add_argument(
        "--repo-root",
        type=Path,
        default=Path("."),
        help="Repo root (default: current directory).",
    )
    parser.add_argument(
        "--manifest",
        type=Path,
        default=Path("data/artifacts/ARTIFACTS_MANIFEST.csv"),
        help="Artifacts manifest CSV (default: data/artifacts/ARTIFACTS_MANIFEST.csv).",
    )
    parser.add_argument(
        "--root",
        type=Path,
        default=Path("data/artifacts"),
        help="Artifacts root directory (default: data/artifacts).",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=Path("data/artifacts/PROVENANCE.local.json"),
        help="Output JSON path (default: data/artifacts/PROVENANCE.local.json).",
    )
    args = parser.parse_args()

    repo_root = args.repo_root.resolve()
    manifest_path = (repo_root / args.manifest).resolve()
    artifacts_root = (repo_root / args.root).resolve()
    out_path = (repo_root / args.output).resolve()

    if not manifest_path.exists():
        raise FileNotFoundError(f"Manifest not found: {manifest_path}")
    if not artifacts_root.exists():
        raise FileNotFoundError(f"Artifacts root not found: {artifacts_root}")

    artifact_paths: list[str] = []
    with manifest_path.open("r", encoding="utf-8", newline="") as f:
        reader = csv.DictReader(f)
        if "artifact_path" not in (reader.fieldnames or []):
            raise ValueError("ARTIFACTS_MANIFEST.csv missing required column: artifact_path")
        for row in reader:
            p = (row.get("artifact_path") or "").strip()
            if p:
                artifact_paths.append(p)

    entries: list[FileHash] = []
    missing: list[str] = []
    for ap in sorted(set(artifact_paths)):
        abs_path = (repo_root / ap).resolve()
        if not abs_path.exists():
            missing.append(ap)
            continue
        if not abs_path.is_file():
            continue
        if artifacts_root not in abs_path.parents and abs_path != artifacts_root:
            raise ValueError(f"Manifest path is outside artifacts root: {ap}")
        rel = abs_path.relative_to(artifacts_root).as_posix()
        st = abs_path.stat()
        entries.append(
            FileHash(
                path=rel,
                size_bytes=st.st_size,
                mtime_utc=iso_utc_from_timestamp(st.st_mtime),
                sha256=sha256_file(abs_path),
            )
        )

    payload = {
        "generated_at_utc": datetime.now(timezone.utc).isoformat(),
        "root": Path("data/artifacts").as_posix(),
        "hashes": [e.__dict__ for e in entries],
        "manifest": Path("data/artifacts/ARTIFACTS_MANIFEST.csv").as_posix(),
        "missing_manifest_paths": sorted(missing),
    }

    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(json.dumps(payload, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    print(f"Wrote: {out_path} ({len(entries)} files; missing={len(missing)})")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/gororoba_cli_data/src/bin/record_artifact_hashes.rs
//! Record local hashes for manifest-tracked artifacts.
//!
//! Migrated from bin/record_artifact_hashes.py.

use anyhow::Result;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::{fs, io::Read, path::Path};

#[derive(Serialize, Deserialize, Debug)]
struct ArtifactProvenance {
    generated_at_utc: DateTime<Utc>,
    root: String,
    hashes: Vec<FileHash>,
    manifest: String,
    missing_manifest_paths: Vec<String>,
}

#[derive(Serialize, Deserialize, Debug)]
struct FileHash {
    path: String,
    size_bytes: u64,
    mtime_utc: DateTime<Utc>,
    sha256: String,
}

fn sha256_file(path: &Path) -> Result<String> {
    let mut file = fs::File::open(path)?;
    let mut hasher = Sha256::new();
    let mut buffer = [0u8; 8192];
    loop {
        let read = file.read(&mut buffer)?;
        if read == 0 {
            break;
        }
        hasher.update(&buffer[..read]);
    }
    Ok(hasher
        .finalize()
        .iter()
        .map(|byte| format!("{byte:02x}"))
        .collect())
}

fn main() -> Result<()> {
    let manifest_path = Path::new("data/artifacts/ARTIFACTS_MANIFEST.csv");
    let _artifacts_root = Path::new("data/artifacts");
    let output_path = Path::new("data/artifacts/PROVENANCE.local.json");

    if !manifest_path.exists() {
        return Ok(()); // Skip if no manifest
    }

    let mut reader = csv::Reader::from_path(manifest_path)?;
    let mut artifact_paths = Vec::new();
    for result in reader.records() {
        let record = result?;
        if let Some(p) = record.get(0)
            && !p.is_empty()
        {
            artifact_paths.push(p.to_string());
        }
    }

    let mut entries = Vec::new();
    let mut missing = Vec::new();

    for ap in artifact_paths {
        let path = Path::new(&ap);
        if !path.exists() {
            missing.push(ap);
            continue;
        }
        if !path.is_file() {
            continue;
        }

        let st = fs::metadata(path)?;
        entries.append(&mut vec![FileHash {
            path: ap.clone(),
            size_bytes: st.len(),
            mtime_utc: st.modified()?.into(),
            sha256: sha256_file(path)?,
        }]);
    }

    let payload = ArtifactProvenance {
        generated_at_utc: Utc::now(),
        root: "data/artifacts".to_string(),
        hashes: entries,
        manifest: manifest_path.to_string_lossy().to_string(),
        missing_manifest_paths: missing,
    };

    let json = serde_json::to_string_pretty(&payload)?;
    fs::write(output_path, json)?;

    println!("Wrote: {}", output_path.display());
    Ok(())
}
